Feature/filestorage sharenetworks list

ehdou opened the pull request:
Add support for Share Network List
For #130

ehdou commented:
[List function](https://github.com/openstack/manila/blob/16c522bf5e5c3d37018caef11dd7852462193327/manila/api/v1/share_networks.py#L108) in Manila code.

ehdou commented:
@jrperritt This PR is ready for review. It's based on [PR#122](https://github.com/gophercloud/gophercloud/pull/122) and add support for list

mvalkon commented:
@ehdou please replace #114 with #130 in the PR description.

jrperritt commented on openstack/sharedfilesystems/v2/sharenetworks/requests.go:
```diff
@@ -50,3 +53,60 @@ func Delete(client *gophercloud.ServiceClient, id string) (r DeleteResult) {
+	// The Share Network description
+	Description string `q:"description"`
+	// The Share Network IP version
+	IPVersion int `q:"ip_version"`
```
s/int/gophercloud.IPVersion

ehdou replied:
fixed

jrperritt commented on openstack/sharedfilesystems/v2/sharenetworks/requests.go:
```diff
@@ -50,3 +53,60 @@ func Delete(client *gophercloud.ServiceClient, id string) (r DeleteResult) {
+	// The Share Network segmentation ID
+	SegmentationID int `q:"segmentation_id"`
+	// List all share networks created after the given date
+	CreatedSince string `q:"created_since"`
```
These should probably be date objects commiserate with the required format.

ehdou replied:
Do you mean using gophercloud.JSONRFC3339MilliNoZ instead of string for the dates ? If so, that doesn't seem to work. BuildQueryString crashes as it is calling isZero() on the value.

jrperritt commented on openstack/sharedfilesystems/v2/sharenetworks/results.go:
```diff
@@ -28,15 +31,36 @@ type ShareNetwork struct {
 
+// ShareNetworkPage is a pagination.pager that is returned from a call to the List function.
+type ShareNetworkPage struct {
+	pagination.SinglePageBase
```
I see limit and offset options: https://github.com/openstack/manila/blob/16c522bf5e5c3d37018caef11dd7852462193327/manila/api/v1/share_networks.py#L172

Are you sure there is no pagination allowed for this type?

jrperritt commented on openstack/sharedfilesystems/v2/sharenetworks/urls.go:
```diff
@@ -9,3 +9,7 @@ func createURL(c *gophercloud.ServiceClient) string {
 }
+
+func listURL(c *gophercloud.ServiceClient) string {
+	return c.ServiceURL("share-networks", "detail")
```
It looks like there are separate `list` and `list detail` operations?

ehdou replied:
There is both operations but I chose to implement only one as we are planning to only use this one. Would you like me to implement the other one, too ?

jrperritt replied:
Not necessarily, but this isn't the `List` URL, right? It's the `List Detail` URL. We should name the operation `ListDetail` (and this function `listDetailURL`) if it's returning the full list response.

ehdou replied:
Done

jrperritt commented on openstack/sharedfilesystems/v2/sharenetworks/results.go:
```diff
@@ -28,15 +31,36 @@ type ShareNetwork struct {
 	Description string `json:"description"`
 	// The date and time stamp when the Share Network was created
-	CreatedAt string `json:"created_at"`
+	CreatedAt gophercloud.JSONRFC3339MilliNoZ `json:"created_at"`
```
:+1:

ehdou commented:
@jrperritt Thanks for your comments, I fixed them and rebased this branch

jrperritt commented on openstack/sharedfilesystems/v2/sharenetworks/results.go:
```diff
@@ -28,15 +31,36 @@ type ShareNetwork struct {
 
+// ShareNetworkPage is a pagination.pager that is returned from a call to the List function.
+type ShareNetworkPage struct {
+	pagination.LinkedPageBase
```
Why is this a `LinkedPageBase` (i.e. where are the links in the page responses)?

ehdou replied:
Sorry I probably misunderstood the pagination types. Which type should I use when I have both `limit` and `offset` as query parameters ? I don't see any that would work and I can't find an example in the other services.

jrperritt replied:
It's probably this: https://github.com/gophercloud/gophercloud/blob/master/pagination/marker.go#L19

For that, you'll need to create a `LastMarker` method and a `NextPageURL` that sets `offset` in the query string instead of `marker`.

ehdou replied:
Done

jrperritt commented on acceptance/openstack/sharedfilesystems/v2/sharenetworks_test.go:
```diff
@@ -21,3 +22,66 @@ func TestShareNetworkCreateDestroy(t *testing.T) {
 	defer DeleteShareNetwork(t, client, shareNetwork)
 }
+
+func TestShareNetworkList(t *testing.T) {
```
I suspect if you had a pagination acceptance test (instead of calling `AllPages`) with more than 1 page, you'd see if the `LinkedPageBase` is correct (I suspect it isn't).

ehdou replied:
I added one now

ehdou commented:
@jrperritt I updated the list function to MarkerPageBase. LastMarker should now return the correct `offset` depending on the values of `offset` and `limit` in the previous query. If pagination is not needed (because `limit` is not present for example), then `offset` is set to maxInt value. I also added an acceptance test for paginaton

jrperritt commented:
+2
